{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Generate QC Report \n",
    "Author: Liam T McKay (liammckay19)\n",
    "\n",
    "This notebook compiles a json file with this information about a sample:\n",
    "    - # of uniquely mapped \n",
    "    - # uniquely mapped non-duplicate \n",
    "    - # uniquely mapped exonic non-duplicate\n",
    "    - # total reads\n",
    "    - % UMEND \n",
    "    - % Duplicates\n",
    "    - RIN score\n",
    "    - the 95th percentile of gene expression\n",
    "    - # expressed genes\n",
    "    - # pan-cancer upoutliers\n",
    "### Input: \n",
    "Notebook 1.json\n",
    "\n",
    "Notebook 4.0.json\n",
    "\n",
    "Log.final.out, fastqc files \n",
    "\n",
    "### Output: (Saves to json file called 4.25.json)\n",
    "\n",
    "\n",
    "`Date`\n",
    "\n",
    "`Threshold Statement`\n",
    "\n",
    "`Sample_id`\n",
    "\n",
    "`Read count table`\n",
    "\n",
    ">`sample_id `\n",
    "\n",
    ">`Total_sequences`\n",
    "\n",
    ">`UM`\n",
    "\n",
    ">`UMND`\n",
    "\n",
    ">`UMEND `\n",
    "\n",
    ">`Multimapped_reads`\n",
    "   \n",
    " `UMEND reads/Total reads`\n",
    ">`Measure, Result, Reference_range, In_reference_range?, n_samples`\n",
    "\n",
    "`Duplicate reads/Total reads`\n",
    ">`Measure, Result, Reference_range, In_reference_range?, n_samples`\n",
    "\n",
    "`RIN`\n",
    ">`Measure, Result, Reference_range, In_reference_range?, n_samples`\n",
    "\n",
    "`Expressed genes (*1000)`\n",
    ">`Measure, Result, Reference_range, In_reference_range?, n_samples`\n",
    "\n",
    "`Pan-cancer up outliers`\n",
    ">`Measure, Result, Reference_range, In_reference_range?, n_samples`\n",
    "\n",
    "`95th percentile of genes in sample (log2(TPM)+1)`\n",
    ">`Measure, Result, Reference_range, In_reference_range?, n_samples`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# global parameters\n",
    "umendThreshold = 1e7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# This cell loads the configuration and contains some helper functions.\n",
    "# Please don't edit it.\n",
    "import os\n",
    "import glob\n",
    "import json\n",
    "import logging\n",
    "import sys\n",
    "import numpy as np\n",
    "import datetime\n",
    "from collections import OrderedDict\n",
    "\n",
    "with open(\"./conf.json\",\"r\") as conf:\n",
    "    c=json.load(conf)\n",
    "sample_id = c[\"sample_id\"]    \n",
    "print(\"Running on sample: {}\".format(sample_id))\n",
    "logging.basicConfig(**c[\"info\"][\"logging_config\"])\n",
    "logging.info(\"4.25: Generate QC Report\")\n",
    "\n",
    "def and_log(s):\n",
    "    logging.info(s)\n",
    "    return s\n",
    "\n",
    "j = {}\n",
    "\n",
    "def load_notebook_output(notebook_num):\n",
    "    outputfile = \"./{}.json\".format(notebook_num)\n",
    "    try:\n",
    "        with open(outputfile, \"r\") as f:\n",
    "           result = json.load(f)\n",
    "        return result\n",
    "    except IOError:\n",
    "        print(\"Error! Couldn't find output of previous notebook at {}\".format(outputfile))\n",
    "        return {}\n",
    "print and_log(\"QC Report Results\\n-------------------------\")\n",
    "# What is the name of your notebook? Fill this out. For example, for \"Template.ipynb\", write \"Template\".\n",
    "notebook_name = \"4.25_qc-reporting\"\n",
    "sample_name = c[\"sample_id\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Requires 1.0 and 4.0 json files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get rsem results (# of expressed genes) and outlier results (pan cancer up outliers)\n",
    "nbOne_rsemResults = load_notebook_output(1)\n",
    "nbFour_outlierResults = load_notebook_output(4.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get redCap file path and QC_reference ranges path\n",
    "redCap_RIN_score_path_and_filename = c[\"ref_file\"][\"redcap_download_rinscore\"]\n",
    "reference_ranges_path = c[\"ref_file\"][\"reference_ranges\"] \n",
    "if not redCap_RIN_score_path_and_filename:\n",
    "    print and_log(\"QC Report Notice: no redCap_RIN_score_path_and_filename in conf.json\")\n",
    "if not reference_ranges_path:\n",
    "    print and_log(\"QC Report Notice: no reference_ranges_path in conf.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def findData(filePath, beginningSearchString, endSearchString):\n",
    "    with open(filePath, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        line = \"\".join(lines)  # join entire html file into one line\n",
    "        start = line.find(beginningSearchString) + len(beginningSearchString)\n",
    "        end = line.find(endSearchString)\n",
    "    return line[start:end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def findFilePaths(newest=True):\n",
    "    \"\"\"\n",
    "    Parse file paths in mustard prism for qc reporting\n",
    "    :param newest: Returns highest version of cgl-pipe, and bam-umend-qc\n",
    "    :return: file paths to fastqc html, log.final.out, bam umend qc json\n",
    "    \"\"\"\n",
    "    # secondary paths\n",
    "    cgl_search = \"/ucsc_cgl-rnaseq-cgl-pipeline-*/\"\n",
    "    \n",
    "    fastqcPath = c['dir']['secondary'] + cgl_search + \"QC/fastQC/R1_fastqc.html\"\n",
    "    logfinaloutPath = c['dir']['secondary'] + cgl_search + \"QC/STAR/Log.final.out\"\n",
    "    \n",
    "    fastqc = glob.glob(os.path.join(fastqcPath))\n",
    "    logfinalout = glob.glob(os.path.join(logfinaloutPath))\n",
    "    \n",
    "    # sort globed paths by highest numbers first \n",
    "    if newest:\n",
    "        fastqc.sort(reverse=True)\n",
    "        logfinalout.sort(reverse=True)\n",
    "    try:\n",
    "        fastqc_Found = fastqc[0]\n",
    "    except IndexError:\n",
    "        print and_log(\"QC Report Notice: R1_fastqc.html file not found. Read from \"+ fastqcPath)\n",
    "        fastqc_Found = \"\"\n",
    "        \n",
    "    try:\n",
    "        logfinalout_Found = logfinalout[0]\n",
    "    except IndexError:\n",
    "        print and_log(\"QC Report Notice: Log.final.out file not found. Read from \"+logfinaloutPath)\n",
    "        logfinalout_Found = \"\"\n",
    "\n",
    "    return fastqc_Found, logfinalout_Found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get read totals for: total sequences, uniquely mapped, multi-mapped reads\n",
    "\n",
    "# find file paths for fastqc, log.final.out, bamumendqc.json\n",
    "fastqcHtmlPath, logFinalPath = findFilePaths(newest=True)\n",
    "\n",
    "# total sequences\n",
    "try:\n",
    "    totalSequences = findData(fastqcHtmlPath, \"Total Sequences</td><td>\",\n",
    "                              \"</td></tr><tr><td>Sequences flagged as poor quality\")\n",
    "    totalSequences = float(totalSequences.rstrip())\n",
    "except IOError:\n",
    "    totalSequences = \"NA\"\n",
    "    print and_log(\"QC Report Notice: total sequences not found in \"+fastqcHtmlPath+\".\\n total_sequences is NA\")\n",
    "\n",
    "# uniquely mapped\n",
    "try:\n",
    "    uniquelyMapped = \"\".join(findData(logFinalPath, \"Uniquely mapped reads number |\",\n",
    "                                  \"Uniquely mapped reads %\").split())  # split with no args to remove all whitespace\n",
    "    uniquelyMapped = float(uniquelyMapped.rstrip())\n",
    "except IOError:\n",
    "    print and_log(\"QC Report Notice: uniquely_mapped not found in \"+logFinalPath+\".\\n uniquely_mapped is NA\")\n",
    "    uniquelyMapped = \"NA\"\n",
    "\n",
    "# multimapped \n",
    "try:\n",
    "    multiMapped = \"\".join(findData(logFinalPath,\"Number of reads mapped to multiple loci |\", \n",
    "                                   \"% of reads mapped to multiple loci\").split())  \n",
    "    # split with no args to remove all whitespace\n",
    "    multiMapped = float(multiMapped.rstrip())\n",
    "except IOError:\n",
    "    print and_log(\"QC Report Notice: multiMapped not found in \"+logFinalPath+\".\\n multiMapped is NA\")\n",
    "    multiMapped = \"NA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get outlier results \n",
    "umend_results = None\n",
    "try:\n",
    "    uniquelyMappedNonDuplicate = str(c['info']['secondary_qc'][\"uniqMappedNonDupeReadCount\"])\n",
    "    umend = str(c['info']['secondary_qc'][\"estExonicUniqMappedNonDupeReadCount\"])\n",
    "    \n",
    "    # remove all carriage returns for output\n",
    "    uniquelyMappedNonDuplicate = float(uniquelyMappedNonDuplicate.rstrip())\n",
    "    umend = float(umend.rstrip())\n",
    "    \n",
    "    pctUMEND = umend / totalSequences\n",
    "except ValueError:\n",
    "    print and_log(\"QC Report Notice: umend_results are NA. \\n\\tUMND, UMEND, pct UMEND, pct Duplicate are NA\")\n",
    "    uniquelyMappedNonDuplicate = 'NA'\n",
    "    umend = 'NA'\n",
    "    pctUMEND = \"NA\"\n",
    "except TypeError:\n",
    "    print and_log(\"QC Report Notice: UMND, UMEND, pct UMEND, pct Duplicate are NA\")\n",
    "    uniquelyMappedNonDuplicate = 'NA'\n",
    "    umend = 'NA'\n",
    "    pctUMEND = \"NA\"\n",
    "    \n",
    "try:\n",
    "    pctDuplicate = 1 - (float(uniquelyMappedNonDuplicate) / float(uniquelyMapped))\n",
    "except ValueError:\n",
    "    print and_log(\"4.25: Notice: uniquelyMappedNonDuplicate and/or uniquelyMapped not found. \"+\n",
    "                    \"\\n\\tUMND, UMEND, pct UMEND, pct Duplicate are NA\")\n",
    "    pctDuplicate = \"NA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get pan cancer up outliers and 95th Percentile of tumor sample\n",
    "try:\n",
    "    pcUpOutliers = 0\n",
    "    for genesResult in nbFour_outlierResults['outlier_results']['pc_outlier'].values():\n",
    "        if genesResult == u'pc_up' and genesResult != \"\":\n",
    "            pcUpOutliers += 1\n",
    "except KeyError:\n",
    "    print and_log(\"QC Report Notice: 4.0.json doesn't have one of [outlier_results][pc_outlier] or neither keys.\"+\n",
    "                  \"\\n\\tPan-cancer Outliers = NA\")\n",
    "    pcUpOutliers = \"NA\"\n",
    "\n",
    "try:\n",
    "    sampleExpressionArray = np.array([i for i in nbFour_outlierResults['outlier_results']['sample'].values()], float)\n",
    "    ninetyFifthPercentile = float(np.percentile(sampleExpressionArray,95))\n",
    "except KeyError:\n",
    "    print and_log(\"QC Report Notice: 4.0.json doesn't have one of [outlier_results][sample] or neither keys.\"+\n",
    "                  \"95th Percentile = NA\")\n",
    "    ninetyFifthPercentile = \"NA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get RIN score for all samples \"redcap_combined_sample_donor_data_downloaded_via_API_*.txt\"\n",
    "if c['ref_file']['redcap_download_rinscore']:\n",
    "    rinData = np.genfromtxt(c['ref_file']['redcap_download_rinscore'], \n",
    "                            dtype=str, delimiter='\\t', usecols=(0, 9), skip_header=1)\n",
    "    try:\n",
    "        sampleRow = np.where(rinData == sample_name)[0][0]\n",
    "        rinScore = rinData[sampleRow][1]\n",
    "        if \"n/a\" not in rinScore:\n",
    "            rinScore = float(rinScore)\n",
    "        else:\n",
    "            rinScore = rinScore.upper().replace(\"/\",\"\")\n",
    "    except (IndexError, ValueError):\n",
    "        rinScore = \"NA\"\n",
    "        print and_log(\"QC Report Notice: redCap_RIN_score_path_and_filename = None \\nrin score not found\")\n",
    "else:\n",
    "    print and_log(\"QC Report Notice: redCap_RIN_score_path_and_filename = None \\nrin score not found\")\n",
    "    rinScore = \"NA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get expressed genes\n",
    "try:\n",
    "    geneExpressionValues = nbOne_rsemResults[\"tpm_hugo_norm_uniq\"][sample_id].values()\n",
    "    nExpressedGenes = np.count_nonzero([float(a) for a in geneExpressionValues]) / 1000.0\n",
    "except KeyError:\n",
    "    print and_log(\"QC ReportNotice: number of expressed genes not found.\\n saving NA\")\n",
    "    nExpressedGenes = \"NA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# get reference ranges\n",
    "# =========== as of coding this the ordering of rows should be ===========\n",
    "# 0 UMEND reads/Total reads\n",
    "# 1 Duplicate reads/Total reads\n",
    "# 2 RIN\n",
    "# 3 Expressed genes (*1000)\n",
    "# 4 Pan-cancer up outliers\n",
    "# 5 95th percentile of genes in sample - (log2(tpm+1))\n",
    "\n",
    "try:\n",
    "    referenceRanges = np.genfromtxt(c['ref_file']['reference_ranges'],\n",
    "                                    dtype=[('Measure', str), ('ref_min', float), ('ref_max', float),\n",
    "                                           ('n_samples', int)], delimiter='\\t', skip_header=0,\n",
    "                                    usecols=(0, 1, 2, 3))\n",
    "except IOError:\n",
    "    referenceRanges = np.array([[1]*5]*7)\n",
    "    print(\"QC Report Notice: reference ranges not found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# generate threshold statement\n",
    "umendThresInMil = str(int(umendThreshold/10**6))\n",
    "try:\n",
    "    if float(umend) > umendThreshold:  # 10 million umend reads threshold\n",
    "        thresholdStatement = \"This sample exceeds the minimum required threshold of \"+umendThresInMil+\\\n",
    "            \" million reads.\"\n",
    "    else:\n",
    "        thresholdStatement = \"This sample does not meet the minimum required threshold of \"+umendThresInMil+\\\n",
    "            \" million reads.\"\n",
    "except ValueError:\n",
    "    thresholdStatement = \"This sample does not meet the minimum required threshold of \"+umendThresInMil+\\\n",
    "        \" million reads.\"\n",
    "print and_log(thresholdStatement)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Today's date - will be printed in Summary+slides\n",
    "today = datetime.datetime.now()\n",
    "generatedTime = \"Generated on \" + str(today.day) + \" \" + today.strftime(\"%B\") + \", \" + str(today.year)\n",
    "logging.info(generatedTime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def updateJsonInfo(measure = \"NA\", rounded_measurement=\"NA\", measurement=\"NA\", refRange=[[1]*5]*7, \n",
    "                   refNum=0):\n",
    "    \"\"\"\n",
    "    Organizes information about the measures associated with the qc_report into a ordered dictionary format\n",
    "    for writing to the final output json. \n",
    "    :param measure: Aspect of the sample that has been measured and gathered here for QC reporting\n",
    "    :param refNum: Row number associated with the measure from the qc_reference ranges tsv file\n",
    "    :param refRange: Actual reference range data in a numpy array in the order of the qc_reference ranges tsv file\n",
    "    :param result: Values of the measure \n",
    "    \"\"\"\n",
    "    need_asterisk = False\n",
    "    reference = [refRange[refNum][1], refRange[refNum][2]]\n",
    "    n_samples = refRange[refNum][3]\n",
    "    added_precision_measurement = rounded_measurement\n",
    "    refYesOrNo = \"No\"\n",
    "    if measure == \"UMEND_reads/Total_reads\":\n",
    "        original_lower_rounded_bound = round(reference[0], 3) # round special case \n",
    "        original_upper_rounded_bound = round(reference[1], 3) # for pc-upoutliers\n",
    "    else:\n",
    "        original_lower_rounded_bound = round(reference[0], 2) # round special case \n",
    "        original_upper_rounded_bound = round(reference[1], 2) # for pc-upoutliers\n",
    "    lower_extended_bound = original_lower_rounded_bound\n",
    "    upper_extended_bound = original_upper_rounded_bound\n",
    "    \n",
    "    if \"NA\" in str(measurement):\n",
    "        refYesOrNo = \"NA\"\n",
    "    else:\n",
    "        if reference[0] <= measurement <= reference[1]:\n",
    "            refYesOrNo = \"Yes\"\n",
    "        elif reference[0] <= measurement and measure == \"RIN\": # only check lower bound ref range for RIN\n",
    "            refYesOrNo = \"Yes\"\n",
    "        if rounded_measurement == original_lower_rounded_bound or rounded_measurement == original_upper_rounded_bound:\n",
    "            # add decimals if rounded measurements are the same\n",
    "            need_asterisk = True\n",
    "            i = 1\n",
    "            while (\n",
    "            float(added_precision_measurement) == float(lower_extended_bound) or \n",
    "            float(added_precision_measurement) == float(upper_extended_bound)):\n",
    "                added_precision_measurement = round(measurement, len(str(rounded_measurement))+i)\n",
    "                lower_extended_bound = round(reference[0], len(str(rounded_measurement))+i)\n",
    "                upper_extended_bound = round(reference[1], len(str(rounded_measurement))+i)\n",
    "                print(measurement,len(str(rounded_measurement))+i)\n",
    "                i += 1\n",
    "                if len(str(rounded_measurement))+i > 6:  # only round to 6 decimal places\n",
    "                    break\n",
    "    \n",
    "    qcInfo = OrderedDict([\n",
    "        (\"Measure\", measure),\n",
    "        (\"Result\", added_precision_measurement),\n",
    "        (\"Rounded_result\", rounded_measurement),\n",
    "        (\"Precise_result\", measurement),\n",
    "        (\"Reference_range\", str(lower_extended_bound) + \" - \" + str(upper_extended_bound)),\n",
    "        (\"Rounded_reference_range\", str(original_lower_rounded_bound) + \" - \" + str(original_upper_rounded_bound)),\n",
    "        (\"Precise_reference_lower\", reference[0]),\n",
    "        (\"Precise_reference_upper\", reference[1]),\n",
    "        (\"In_reference_range?\", refYesOrNo),\n",
    "        (\"n_samples\", n_samples),\n",
    "        (\"asterisk\", need_asterisk)]\n",
    "    )\n",
    "    \n",
    "    return qcInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# create qc report\n",
    "qcReport = OrderedDict([\n",
    "    (\"Date\",generatedTime),\n",
    "    (\"Threshold_Statement\",thresholdStatement),\n",
    "    (\"UMEND_threshold\",umendThreshold),\n",
    "    (\"Sample_id\",sample_name),\n",
    "    ('Read_count_table',''),\n",
    "    (\"UMEND_reads/Total_reads\",''),\n",
    "    (\"Duplicate_reads/Total_reads\",''),\n",
    "    (\"RIN\",''),\n",
    "    (\"Expressed_genes_(*1000)\",''),\n",
    "    (\"Pan-cancer_up_outliers\",''),\n",
    "    (\"95th_percentile_of_genes_in_sample_(log2(TPM)+1)\",'')\n",
    "])\n",
    "\n",
    "umendThresholdAsterisk = False\n",
    "if \"NA\" not in str(umend):\n",
    "    if round(umend/10**6,2) == round(umendThreshold,2):\n",
    "        umendThresholdAsterisk = True\n",
    "\n",
    "qcReport[\"Read_count_table\"] = OrderedDict([\n",
    "        (\"sample_id\", sample_name),\n",
    "        (\"Total_sequences\", totalSequences),\n",
    "        (\"UM\", uniquelyMapped),\n",
    "        (\"UMND\", uniquelyMappedNonDuplicate),\n",
    "        (\"UMEND\", umend),\n",
    "        (\"Multimapped_reads\", multiMapped),\n",
    "        (\"asterisk\", umendThresholdAsterisk)\n",
    "    ])\n",
    "\n",
    "# Measure name, full precision, rounded precision. \n",
    "# -- checks to make sure the values are float before rounding\n",
    "\n",
    "\n",
    "measuresForReference = [\n",
    "    (\"UMEND_reads/Total_reads\", (round(pctUMEND, 3) if \"NA\" not in str(pctUMEND) else pctUMEND), \n",
    "     pctUMEND),\n",
    "    \n",
    "    (\"Duplicate_reads/Total_reads\", (round(pctDuplicate, 2) if \"NA\" not in str(pctDuplicate) else pctDuplicate),\n",
    "     pctDuplicate),\n",
    "    \n",
    "    (\"RIN\", (round(rinScore, 2) if \"NA\" not in str(rinScore) else rinScore), rinScore),\n",
    "    \n",
    "    (\"Expressed_genes_(*1000)\", (round(nExpressedGenes, 2) if \"NA\" not in str(nExpressedGenes) else nExpressedGenes),\n",
    "     nExpressedGenes),\n",
    "    \n",
    "    (\"Pan-cancer_up_outliers\", pcUpOutliers,\n",
    "     pcUpOutliers),\n",
    "    \n",
    "    (\"95th_percentile_of_genes_in_sample_(log2(TPM)+1)\",\n",
    "     (round(ninetyFifthPercentile, 2) if \"NA\" not in str(ninetyFifthPercentile) else ninetyFifthPercentile),\n",
    "     ninetyFifthPercentile)\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "for i, measure in enumerate(measuresForReference, 1):\n",
    "    qcReport[measure[0]] = updateJsonInfo(measure=measure[0], \n",
    "                                          rounded_measurement=measure[1], \n",
    "                                          measurement=measure[2], \n",
    "                                          refRange=referenceRanges, \n",
    "                                          refNum=i)\n",
    "\n",
    "with open(c[\"json\"][\"4.25\"], 'w') as outFile:\n",
    "    json.dump(qcReport, outFile, indent=2)\n",
    "\n",
    "print \"Done. Output location: \" + c[\"json\"][\"4.25\"]\n",
    "print(json.dumps(qcReport, indent=4)) # prints json below"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
