{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Generate Isoform report for Summary.html #212\n",
    "Wrapper for isoform_report.R . Based on version:\n",
    "https://github.com/UCSC-Treehouse/isoform_report/blob/ca339b96e0a9dda1281b0b6b064af29dfd4a3e70/isoform_report.R with slight modification.\n",
    "\n",
    "Dependencies - the following R libraries are required to run this script:\n",
    " - dplyr >= 0.8.0\n",
    " - tidyverse >= 1.3.0\n",
    " - treemapify (2.5.3)\n",
    " - Sushi (1.12.0)\n",
    "\n",
    "Sub-dependency notes\n",
    " - Sushi requires BiocManager (1.30.10) to install\n",
    " - tidyverse 1.3.0 has a dependency vctrs; as of 2020/01/15 the latest vctrs (0.2.1) isn't installing, so vctrs 0.2.0 is used instead.\n",
    "\n",
    "Outputs:\n",
    "Output files are placed into an isoform-report subdirectory. For each gene:\n",
    " - {gene}\\_isoform_expression.tsv\n",
    " - Expressed\\_{gene}\\_isoforms\\_in\\_{sample_id}.pdf\n",
    " - Expressed\\_{gene}\\_isoforms\\_in\\_{sample_id}.png\n",
    " - Frequency\\_of\\_{gene}\\_transcript\\_biotypes\\_in\\_{sample_id}.png\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import json\n",
    "import logging\n",
    "import base64\n",
    "import pandas as pd\n",
    "from PIL import Image\n",
    "from string import Template\n",
    "from collections import OrderedDict\n",
    "from distutils.version import LooseVersion\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "with open(\"conf.json\",\"r\") as conf:\n",
    "    c=json.load(conf)\n",
    "sample_id = c[\"sample_id\"]    \n",
    "print(\"Running on sample: {}\".format(sample_id))\n",
    "logging.basicConfig(**c[\"info\"][\"logging_config\"])\n",
    "\n",
    "def and_log(s):\n",
    "    logging.info(s)\n",
    "    return s\n",
    "\n",
    "j = {}\n",
    "\n",
    "def load_notebook_output(notebook_num):\n",
    "    outputfile = \"{}.json\".format(notebook_num)\n",
    "    try:\n",
    "        with open(outputfile, \"r\") as f:\n",
    "            result = json.load(f)\n",
    "        return result\n",
    "    except IOError:\n",
    "        print(\"Error! Couldn't find output of previous notebook at {}\".format(outputfile))\n",
    "        return {}\n",
    "    \n",
    "print and_log(\"8.25: Isoform Report\")\n",
    "notebook_name = \"8.25\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Locate rsem-isoforms.results\n",
    "\n",
    "# pipeline_name = glob-able basedir within secondary (format = PIPELINE-NAME-*)\n",
    "# internal_path = path to file within that pipeline name\n",
    "def locate_in_secondary(pipeline_name, internal_path):\n",
    "    all_pipeline_versions = sorted(glob.glob(os.path.join(c[\"dir\"][\"secondary\"], pipeline_name)),\n",
    "                        key=LooseVersion)\n",
    "    # Get the file, or most recent if there are more than one.\n",
    "    if len(all_pipeline_versions) >= 1:\n",
    "        return os.path.join(all_pipeline_versions[-1], internal_path)\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "rsem_isoforms_path = locate_in_secondary(\n",
    "    \"ucsc_cgl-rnaseq-cgl-pipeline-*\", os.path.join(\"RSEM\", \"rsem_isoforms.results\"))\n",
    "j[\"found_isoforms_file\"] = os.access(rsem_isoforms_path, os.R_OK) # we'll only expect results if file can be read\n",
    "\n",
    "# get genes of interest as comma-separated\n",
    "nb8_results = load_notebook_output(\"8\")\n",
    "leads = nb8_results[\"automated_leads_identified\"]\n",
    "genes_of_interest_list = { k:v for (k, v) in leads[\"results\"].items() if \n",
    "                              leads[\"assay\"][k] == \"druggableUpOutlier\"}.values()\n",
    "genes_of_interest = \",\".join(genes_of_interest_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Parameters for Rscript\n",
    "sid = sample_id\n",
    "genes = genes_of_interest\n",
    "enshugo = c[\"ref_file\"][\"ensembl_hugo_mapping_file\"]\n",
    "rir = rsem_isoforms_path\n",
    "gtf = c[\"ref_file\"][\"treehouse_druggable_gencode\"] # Treehouse_druggable_gencode.v23.annotation.gtf.gz\n",
    "outdir = c[\"dir\"][\"isoform_report_plots_dir\"] # isoform-report\n",
    "\n",
    "print(\"Making isoform report for genes {}\".format(genes_of_interest))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# make output dir\n",
    "try: \n",
    "    os.makedirs(outdir)\n",
    "    print(\"Made output dir {}\".format(outdir))\n",
    "except OSError as e:\n",
    "    print(\"Found error, but perhaps the dir simply already exists?\\nError: {}\".format(e))\n",
    "    if not os.path.isdir(outdir):\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "source": [
    "Run the Rscript. In the case where there is no rsem_isoforms file, it will fail to generate output files, which will be caught in the following steps.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "%%script Rscript - \"$sid\" \"$genes\" \"$enshugo\" \"$rir\" \"$gtf\" \"$outdir\"\n",
    "\n",
    "params <-\n",
    "list(date = structure(18249, class = \"Date\"))\n",
    "\n",
    "\n",
    "library(tidyverse)\n",
    "library(Sushi)\n",
    "library(knitr)\n",
    "library(treemapify)\n",
    "library(RColorBrewer)\n",
    "\n",
    "\n",
    "f_RMD <- isTRUE(getOption('knitr.in.progress')) | interactive()\n",
    "\n",
    "# parameters when run as a script\n",
    "\n",
    "if (! f_RMD) {\n",
    "args<-commandArgs(TRUE)\n",
    "\n",
    "sample_of_interest <- args[1]\n",
    "genes_of_interest_hugo <- strsplit(args[2], \",\") %>% unlist\n",
    "EnsGeneID_Hugo_Observed_Conversions_file <- args[3] # \"EnsGeneID_Hugo_Observed_Conversions.txt\"\n",
    "rsem_isoforms.results_file <- args[4] # \"rsem_isoforms.results\"\n",
    "gtf_file <- args[5] # \"Treehouse_druggable_gencode.v23.annotation.gtf.gz\"\n",
    "output_dir <-  args[6] # \"isoform-report\"\n",
    "}\n",
    "\n",
    "\n",
    "# parameters when run interactively or knitted\n",
    "\n",
    "if (f_RMD) {\n",
    "  genes_of_interest_hugo <- c(\"KIT\", \"PDGFRA\")\n",
    "  sample_of_interest <- \"TH34_1349_S02\"\n",
    "  EnsGeneID_Hugo_Observed_Conversions_file <- \"EnsGeneID_Hugo_Observed_Conversions.txt\"\n",
    "  rsem_isoforms.results_file <- \"rsem_isoforms.results\"\n",
    "  gtf_file <- \"Treehouse_druggable_gencode.v23.annotation.gtf.gz\"\n",
    "  output_dir <- \"isoform-report\"\n",
    "}\n",
    "\n",
    "\n",
    "# if (f_RMD) {\n",
    "print(paste(genes_of_interest_hugo, collapse = \"-\"))\n",
    "print(sample_of_interest)\n",
    "# }\n",
    "\n",
    "\n",
    "ens_hugo_conversions <- read_tsv(EnsGeneID_Hugo_Observed_Conversions_file) %>%  na.omit\n",
    "\n",
    "genes_of_interest_ensembl <- ens_hugo_conversions$EnsGeneID[\n",
    "  ens_hugo_conversions$HugoID %in% genes_of_interest_hugo]\n",
    "\n",
    "iso_results <- read_tsv(rsem_isoforms.results_file)\n",
    "\n",
    "these_genes_iso_results <- iso_results %>% filter(gene_id %in% genes_of_interest_ensembl)\n",
    "\n",
    "gtf_colnames <- c(\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\")\n",
    "\n",
    "# gtf_file <- \"gencode.v23.annotation.gtf.gz\"\n",
    "# gtf_file <- \"Treehouse_druggable_gencode.v23.annotation.gtf.gz\"\n",
    "\n",
    "gencode_v23 <- read_tsv(gtf_file, comment = \"#\", col_names = gtf_colnames)\n",
    "\n",
    "gencode_v23_these_genes <- gencode_v23 %>%\n",
    "  mutate(gene_id = gsub(\"\\\".*$\", \"\", \n",
    "                        gsub(\"^gene_id \\\"\", \"\\\\1\", attribute)),\n",
    "         transcript_id = gsub(\"^.*transcript_id \\\"([A-Z0-9\\\\.]*)\\\".*$\", \n",
    "                              \"\\\\1\", attribute),\n",
    "         feature_length = end - start\n",
    "  )  %>%\n",
    "  filter(gene_id %in% genes_of_interest_ensembl)\n",
    "\n",
    "\n",
    "gencode_v23_these_genes_transcripts <- gencode_v23_these_genes %>%\n",
    "  filter(feature == \"transcript\")\n",
    "\n",
    "KVsep <- fixed(\"; \")  #key-value separator\n",
    "Vsep <- fixed(\" \")     #value separator\n",
    "\n",
    "gencode_v23_these_genes_transcript_minutia <-  gencode_v23_these_genes_transcripts %>% \n",
    "  mutate(KVpairs = str_split(attribute, KVsep)) %>%\n",
    "  unnest(KVpairs) %>%\n",
    "  separate(KVpairs, into = c(\"key\", \"value\"), Vsep) %>%\n",
    "  filter( !(key == \"tag\" & value !=\"basic\"))  %>% # keep tag only if basic\n",
    "  filter(! key %in% c(\"transcript_id\", \"gene_id\")) %>% # these value were already extracted\n",
    "  mutate(value = gsub(\"\\\"\", \"\", value)) %>%\n",
    "  spread(key, value)\n",
    "\n",
    "\n",
    "these_genes_iso_results_anno <- these_genes_iso_results %>%\n",
    "  left_join(gencode_v23_these_genes_transcript_minutia %>% \n",
    "             dplyr::select (-gene_id), \n",
    "            by=\"transcript_id\") %>%\n",
    "  mutate(transcript_id = fct_reorder(transcript_id, IsoPct))\n",
    "\n",
    "n_transcripts_to_analyze <- 10\n",
    "top_iso_results_anno <- these_genes_iso_results_anno %>%\n",
    "  top_n(n_transcripts_to_analyze, IsoPct)\n",
    "  \n",
    "\n",
    "\n",
    "\n",
    "# Error in .f(.x[[i]], ...) : object 'transcript_type' not found\n",
    "# Calls: %>% ... <Anonymous> -> vars_select_eval -> map_if -> map -> .f\n",
    "#Execution halted\n",
    "\n",
    "exon_locations <- gencode_v23_these_genes %>%\n",
    "  filter(feature %in% c(\"exon\", \"UTR\")) %>%\n",
    "  left_join(top_iso_results_anno %>% \n",
    "              dplyr::select(transcript_id, IsoPct, TPM, transcript_type, transcript_name, hugo_id = gene_name),\n",
    "            by=c(\"transcript_id\")) %>%\n",
    "  mutate(score = IsoPct,\n",
    "         transcript_label = paste0(transcript_name, \" (\",\n",
    "                                   #IsoPct, \"%, \", transcript_type, \")\")) %>%\n",
    "                                   IsoPct, \"%)\")) %>%\n",
    "  dplyr::select(chrom = seqname, start, stop = end, \n",
    "                gene = transcript_id, score, \n",
    "                strand, type = feature, IsoPct, TPM, transcript_label, transcript_name, hugo_id) %>%\n",
    "  arrange(desc(IsoPct))\n",
    "\n",
    "\n",
    "\n",
    "plot_gene <- function(submitted_bed_data, buffer_size = 5e2, plot_title = \"\"){\n",
    "  bed_data = data.frame(submitted_bed_data)\n",
    "  chrom <- bed_data$chrom[1]\n",
    "  chromstart = min(bed_data$start) - buffer_size\n",
    "  chromend = max(bed_data$stop) + buffer_size\n",
    "    \n",
    "  # Only colorby if there is more than one unique score or it will crash\n",
    "  if( length(unique(bed_data$score))==1 ){\n",
    "        pg = plotGenes(bed_data,chrom,chromstart,chromend,\n",
    "                 #colorby=log10(bed_data$score+0.001),\n",
    "                 #colorby=bed_data$score,\n",
    "                 #colorbycol= SushiColors(5),colorbyrange=c(0,1.0),\n",
    "                 labeltext=TRUE,maxrows=50,height=0.4,plotgenetype=\"box\",\n",
    "                 packrow = FALSE\n",
    "      )   \n",
    "  } else {\n",
    "        pg = plotGenes(bed_data,chrom,chromstart,chromend,\n",
    "                 #colorby=log10(bed_data$score+0.001),\n",
    "                 colorby=bed_data$score,\n",
    "                 #colorbycol= SushiColors(5),colorbyrange=c(0,1.0),\n",
    "                 labeltext=TRUE,maxrows=50,height=0.4,plotgenetype=\"box\",\n",
    "                 packrow = FALSE\n",
    "      )   \n",
    "  }\n",
    "\n",
    "  labelgenome( chrom, chromstart,chromend,n=3,scale=\"Mb\")\n",
    " # note: add legend has to be hand-placed for each plot, so I've omitted it here\n",
    "  title(main = plot_title, sub = \"Colored by isoform percent, also reported in label\")\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "multi_plot_gene <- function(bed_data) {\n",
    "  # bed_data <- t5\n",
    "  this_title <- paste(\"Expressed\", bed_data$hugo_id[1], \"isoforms in\", sample_of_interest)\n",
    "  \n",
    "  base_filename <- gsub(\" \", \"_\", this_title)\n",
    "\n",
    "  ## If using RMD to generate html output, make plot\n",
    "  if( f_RMD ) plot_gene (bed_data, plot_title = this_title)\n",
    "  \n",
    "  ## If scripted, make plots in output files\n",
    "  if( ! f_RMD ) {\n",
    "    ## Make PDF (small file size, can be endlessly enlarged, inconvenient to embed in html)\n",
    "    pdf(file = file.path(output_dir, paste0(base_filename, \".pdf\")),\n",
    "        width = 8, height = 4)\n",
    "    \n",
    "    plot_gene (bed_data, plot_title = this_title)\n",
    "    \n",
    "    dev.off()  \n",
    "    \n",
    "    ## Make high res PNG (large file size, convenient to embed in html)\n",
    "    png(file = file.path(output_dir, paste0(base_filename, \".png\")),\n",
    "        width = 12, height = 6, units = \"in\", res = 600)\n",
    "    \n",
    "    plot_gene (bed_data, plot_title = this_title)\n",
    "    \n",
    "    dev.off()  \n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "expressed_transcripts <- exon_locations %>% \n",
    "  mutate(gene = transcript_label) %>% \n",
    "  dplyr::filter(TPM > 0) \n",
    "\n",
    "expressed_transcripts %>% \n",
    "  group_by(hugo_id) %>%\n",
    "  group_split %>%\n",
    "  lapply(multi_plot_gene)\n",
    "\n",
    "\n",
    "\n",
    "transcript_biotypes <- c(\"protein_coding\", \"processed_transcript\", \"retained_intron\", \"processed_pseudogene\", \"nonsense_mediated_decay\", \"transcribed_processed_pseudogene\")\n",
    "# biotype_color_codes <- tibble(transcript_biotypes, brewer.pal(12, \"Set1\")[1:length(transcript_biotypes)])\n",
    "biotype_color_codes <- tibble(transcript_biotype = transcript_biotypes, color_code = brewer.pal(length(transcript_biotypes), \"Set1\"))\n",
    "\n",
    "\n",
    " \n",
    "transcript_biotype_colors <- biotype_color_codes$color_code\n",
    "names(transcript_biotype_colors) <-  biotype_color_codes$transcript_biotype\n",
    "\n",
    "\n",
    "\n",
    "plot_biotype_frequency <- function(this_gene_iso_results_anno){\n",
    "  \n",
    "  biotype_freq <- this_gene_iso_results_anno %>% \n",
    "    group_by(transcript_type) %>%\n",
    "    summarize(total_isoform_pct_for_type = sum(IsoPct)) %>%\n",
    "    mutate(biotype_label = paste0(transcript_type, \" (\", total_isoform_pct_for_type, \"%)\"))\n",
    "  \n",
    "  this_title <- paste(\"Frequency of\", this_gene_iso_results_anno$gene_name[1] ,\"transcript biotypes in\", sample_of_interest)\n",
    "  base_filename <- gsub(\" \", \"_\", this_title)\n",
    "  \n",
    "  ggplot(biotype_freq, \n",
    "         aes(fill = transcript_type, \n",
    "             area = total_isoform_pct_for_type, \n",
    "             label = biotype_label)) +\n",
    "    geom_treemap() + \n",
    "    geom_treemap_text(colour = \"white\", \n",
    "                      place = \"centre\") +\n",
    "    labs(title = this_title) +\n",
    "    scale_fill_manual(values = transcript_biotype_colors)\n",
    "  \n",
    "  ggsave(file.path(output_dir, paste0(base_filename, \".png\")), width=5, height = 5)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "these_genes_iso_results_anno %>% \n",
    "  group_by(gene_name) %>%\n",
    "  group_split %>%\n",
    "  lapply(plot_biotype_frequency)\n",
    "\n",
    "\n",
    "\n",
    "output_table <- these_genes_iso_results_anno  %>% \n",
    "  dplyr::filter(TPM > 0) %>%\n",
    "  mutate(log2TPM1 = round(log2(TPM +1),2)) %>%\n",
    "  dplyr::select(transcript_name, length, log2TPM1, strand, IsoPct, transcript_type, transcript_id, gene_name) %>%\n",
    "  arrange(desc(IsoPct)) %>% \n",
    "  group_by(gene_name)\n",
    "\n",
    "if(f_RMD){\n",
    "  list_of_output_tables <- output_table %>% \n",
    "    group_split\n",
    "  \n",
    "  for(i in list_of_output_tables) {\n",
    "    print(kable(x = i))\n",
    "  }\n",
    "} \n",
    "\n",
    "for_silent_output <- output_table %>% \n",
    "  group_split %>%\n",
    "  lapply(function(x) {write_tsv(x, file.path(output_dir, paste0(x$gene_name[1], \"_isoform_expression.tsv\")))})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def image_to_json(path, filename):\n",
    "    try:\n",
    "        with open(os.path.join(path,filename), \"rb\") as f:\n",
    "            return base64.b64encode(f.read())\n",
    "    except IOError:\n",
    "        print(\"Couldn't read {}/{}; skipping\".format(path, filename))\n",
    "        return False\n",
    "        \n",
    "def downscale_png(png_path):\n",
    "    try:\n",
    "        png_to_resize = Image.open(png_path)\n",
    "        png_to_resize.thumbnail((1080,540))\n",
    "        png_to_resize.save(png_path, \"PNG\")\n",
    "        return True\n",
    "    except IOError:\n",
    "        print(\"Couldn't resize image at {}; skipping\".format(png_path))\n",
    "        return False\n",
    "\n",
    "expressed_isoforms_png = Template(\"Expressed_${gene}_isoforms_in_${sample_id}.png\")\n",
    "# We also generate Expressed_${gene}_isoforms_in_${sample_id}.pdf\"\n",
    "frequency_png = Template(\"Frequency_of_${gene}_transcript_biotypes_in_${sample_id}.png\")\n",
    "expression_tsv = Template(\"${gene}_isoform_expression.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Downscale expressed isoform PNGs and import PNGs and table into JSON file\n",
    "# any file not found will be imported as False    \n",
    "\n",
    "j[\"isoform_results\"] = {}\n",
    "for gene in genes_of_interest_list:\n",
    "    j[\"isoform_results\"][gene]={}\n",
    "    png_path = os.path.join(outdir, \n",
    "                            expressed_isoforms_png.substitute(gene=gene, sample_id=sample_id))    \n",
    "    downscale_png(png_path)\n",
    "    # Load images into JSON - will be False if file not found\n",
    "    j[\"isoform_results\"][gene][\"expressed_isoforms_img_data\"]=image_to_json(outdir,\n",
    "        expressed_isoforms_png.substitute(gene=gene, sample_id=sample_id))\n",
    "    j[\"isoform_results\"][gene][\"transcript_biotypes_img_data\"]=image_to_json(outdir,\n",
    "        frequency_png.substitute(gene=gene, sample_id=sample_id))\n",
    "    \n",
    "    try:\n",
    "        isoform_table_file=os.path.join(outdir,expression_tsv.substitute(gene=gene))\n",
    "        isoform_table_json=json.loads(\n",
    "            pd.read_csv(isoform_table_file, delimiter=\"\\t\", dtype=\"str\", na_filter=False\n",
    "                   ).to_json(orient=\"records\"),object_pairs_hook=OrderedDict)\n",
    "    except IOError:\n",
    "        print(\"Couldn't read {}; skipping\".format(isoform_table_file))\n",
    "        isoform_table_json = False\n",
    "    j[\"isoform_results\"][gene][\"isoform_table\"]=isoform_table_json\n",
    "\n",
    "# Store order of isoform keys for proper display in summary\n",
    "j[\"isoform_table_key_order\"]=[\n",
    "    \"transcript_name\",\n",
    "    \"length\",\n",
    "    \"log2TPM1\",\n",
    "    \"strand\",\n",
    "    \"IsoPct\",\n",
    "    \"transcript_type\",\n",
    "    \"transcript_id\",\n",
    "    \"gene_name\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "with open(\"{}.json\".format(notebook_name), \"w\") as jsonfile:\n",
    "    json.dump(j, jsonfile, indent=2)\n",
    "    \n",
    "print(\"Done.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
